File: lib/screens/blackjack_screen.dart
import 'package:flutter/material.dart';
import '../theme.dart';
import '../logic/card_deck.dart';
import '../db/auth_service.dart';
import '../services/sound_service.dart';
import 'config_screen.dart';
import 'auth/login_screen.dart';
import 'transactions_screen.dart';

class BlackjackScreen extends StatefulWidget {
  const BlackjackScreen({Key? key}) : super(key: key);

  @override
  State<BlackjackScreen> createState() => _BlackjackScreenState();
}

class _BlackjackScreenState extends State<BlackjackScreen> {
  final Deck _deck = Deck();
  List<PlayingCard> _playerH = [];
  List<PlayingCard> _splitH = []; // New split hand
  List<PlayingCard> _dealerH = [];
  
  bool _isPlaying = false;
  bool _gameOver = false;
  String _message = "Place your bet.";
  int _betAmount = 100;
  int _activeBet = 0;
  int _splitBet = 0; // Bet for split hand
  int _balance = 0;
  
  int _activeHand = 1; // 1 or 2

  @override
  void initState() {
    super.initState();
    _balance = AuthService.currentBalance;
  }

  void _updateBal(int amount) {
    AuthService.updateBalance(amount, gameName: 'Blackjack');
    setState(() { _balance = AuthService.currentBalance; });
  }

  int _calcHand(List<PlayingCard> hand) {
    int v = 0;
    int a = 0;
    for (var c in hand) {
      v += c.blackjackValue;
      if (c.rank == Rank.ace) a++;
    }
    while (v > 21 && a > 0) {
      v -= 10;
      a--;
    }
    return v;
  }

  void _placeBet() {
    if (_balance < _betAmount) {
      setState(() => _message = "Not enough funds.");
      return;
    }
    _updateBal(-_betAmount);
    SoundService.playClick();
    SoundService.playCardDraw();
    setState(() {
      _activeBet = _betAmount;
      _splitBet = 0;
      _playerH = [_deck.draw(), _deck.draw()];
      _splitH = [];
      _dealerH = [_deck.draw(), _deck.draw()];
      _isPlaying = true;
      _gameOver = false;
      _activeHand = 1;
      _message = "";
      if (_calcHand(_playerH) == 21) {
        _endGame();
      }
    });
  }

  void _fold() { // Surrender
    if (_playerH.length == 2 && !_gameOver && _splitH.isEmpty) {
      setState(() {
        _message = "SURRENDER. RETURN HALF BET.";
        _updateBal((_activeBet / 2).toInt());
        _gameOver = true;
        _isPlaying = false;
      });
    }
  }

  void _doubleDown() {
    if ((_activeHand == 1 ? _playerH.length : _splitH.length) == 2 && !_gameOver) {
      if (_balance >= (_activeHand == 1 ? _activeBet : _splitBet)) {
        int bet = _activeHand == 1 ? _activeBet : _splitBet;
        _updateBal(-bet);
        if (_activeHand == 1) _activeBet *= 2; else _splitBet *= 2;
        SoundService.playCardDraw();
        setState(() {
          if (_activeHand == 1) {
             _playerH.add(_deck.draw());
             if (_calcHand(_playerH) > 21) {
               if (_splitH.isEmpty) { _message = "BUST! YOU LOSE."; _endGame(); } else { _stand(); }
             } else { _stand(); }
          } else {
             _splitH.add(_deck.draw());
             _stand();
          }
        });
      } else {
        ScaffoldMessenger.of(context).showSnackBar(const SnackBar(content: Text("Not enough funds to Double.", style: TextStyle(color: Colors.white)), backgroundColor: Colors.red));
      }
    }
  }

  void _split() {
    if (_playerH.length == 2 && _splitH.isEmpty && _playerH[0].blackjackValue == _playerH[1].blackjackValue) {
      if (_balance >= _activeBet) {
        _updateBal(-_activeBet);
        SoundService.playCardDraw();
        setState(() {
          _splitBet = _activeBet;
          _splitH = [_playerH.removeLast(), _deck.draw()];
          _playerH.add(_deck.draw());
        });
      } else {
        ScaffoldMessenger.of(context).showSnackBar(const SnackBar(content: Text("Not enough funds to Split.", style: TextStyle(color: Colors.white)), backgroundColor: Colors.red));
      }
    }
  }

  void _hit() {
    SoundService.playCardDraw();
    setState(() {
      if (_activeHand == 1) {
        _playerH.add(_deck.draw());
        if (_calcHand(_playerH) > 21) {
           if (_splitH.isNotEmpty) {
             _activeHand = 2; // Move to hand 2
           } else {
             _message = "BUST! YOU LOSE.";
             _endGame();
           }
        }
      } else {
        _splitH.add(_deck.draw());
        if (_calcHand(_splitH) > 21) {
           _stand();
        }
      }
    });
  }

  void _stand() {
    setState(() {
      if (_splitH.isNotEmpty && _activeHand == 1) {
        _activeHand = 2; // Move to second hand
      } else {
        while (_calcHand(_dealerH) < 17) {
          _dealerH.add(_deck.draw());
        }
        _endGame();
      }
    });
  }

  void _endGame() {
    _gameOver = true;
    _isPlaying = false;
    
    int dV = _calcHand(_dealerH);
    int pV = _calcHand(_playerH);
    
    // Evaluate Hand 1
    String res1 = "";
    if (pV <= 21) {
      if (pV == 21 && _playerH.length == 2 && !(dV == 21 && _dealerH.length == 2) && _splitH.isEmpty) {
        res1 = "BLACKJACK! (3:2)";
        SoundService.playWin();
        _updateBal((_activeBet * 2.5).toInt());
      } else if (dV > 21 || pV > dV) {
        res1 = "WIN";
        SoundService.playWin();
        _updateBal(_activeBet * 2);
      } else if (pV < dV) {
        res1 = "LOSE";
      } else {
        res1 = "PUSH";
        _updateBal(_activeBet);
      }
    } else {
      res1 = "BUST";
    }

    // Evaluate Hand 2 (if exists)
    if (_splitH.isNotEmpty) {
      int sV = _calcHand(_splitH);
      String res2 = "";
      if (sV <= 21) {
         if (dV > 21 || sV > dV) {
            res2 = "WIN";
            SoundService.playWin();
            _updateBal(_splitBet * 2);
         } else if (sV < dV) {
            res2 = "LOSE";
         } else {
            res2 = "PUSH";
            _updateBal(_splitBet);
         }
      } else {
        res2 = "BUST";
      }
      _message = "H1: $res1 | H2: $res2";
    } else {
      _message = res1 == "WIN" ? "YOU WIN! +\$${_activeBet * 2}" : res1 == "LOSE" ? "DEALER WINS." : res1 == "PUSH" ? "PUSH (TIE)." : res1 == "BUST" ? "BUST! YOU LOSE." : res1;
    }
  }

  Widget _buildChip(int amount) {
    return InkWell(
      onTap: () {
        if (_betAmount + amount <= _balance) {
          SoundService.playClick();
          setState(() => _betAmount += amount);
        }
      },
      child: Container(
        width: 50, height: 50,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: AppColors.surfaceContainerHigh,
          border: Border.all(color: AppColors.primary, width: 2),
        ),
        alignment: Alignment.center,
        child: Text("\$$amount", style: const TextStyle(color: AppColors.primary, fontWeight: FontWeight.bold)),
      ),
    );
  }

  Widget _buildActionButton(String label, VoidCallback? onPressed, {bool isPrimary = false}) {
    return InkWell(
      onTap: onPressed ?? () {
        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
          content: Text("$label action coming soon!"),
          backgroundColor: AppColors.surfaceVariant,
        ));
      },
      child: Container(
        width: 60,
        height: 60,
        decoration: BoxDecoration(
          color: isPrimary ? AppColors.primary : Colors.black.withValues(alpha: 0.8),
          border: Border.all(color: AppColors.primary, width: isPrimary ? 0 : 1),
          borderRadius: BorderRadius.circular(4),
        ),
        alignment: Alignment.center,
        child: Text(
          label,
          style: TextStyle(
            color: isPrimary ? Colors.black : AppColors.primary,
            fontSize: 10,
            fontWeight: FontWeight.bold,
            letterSpacing: 1.0,
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      appBar: AppBar(
        backgroundColor: Colors.transparent,
        elevation: 0,
        title: FittedBox(
          fit: BoxFit.scaleDown,
          child: Column(
            children: [
              Text("\$$_balance", style: Theme.of(context).textTheme.titleLarge?.copyWith(color: AppColors.primary, fontWeight: FontWeight.bold)),
              const Text("AVAILABLE BALANCE", style: TextStyle(color: AppColors.onSurfaceVariant, fontSize: 8, letterSpacing: 1.5, fontWeight: FontWeight.bold)),
            ],
          ),
        ),
        centerTitle: true,
        actions: [
          PopupMenuButton<String>(
            icon: const Icon(Icons.menu, color: AppColors.primary),
            color: AppColors.surfaceContainerHigh,
            onSelected: (value) {
              if (value == 'profile') {
                Navigator.push(context, MaterialPageRoute(builder: (_) => const ConfigScreen()));
              } else if (value == 'logout') {
                AuthService.logout();
                Navigator.pushAndRemoveUntil(context, MaterialPageRoute(builder: (_) => const LoginScreen()), (r) => false);
              }
            },
            itemBuilder: (context) => [
              const PopupMenuItem(
                value: 'profile',
                child: Text('Profile & Settings', style: TextStyle(color: AppColors.primary)),
              ),
              const PopupMenuItem(
                value: 'logout',
                child: Text('Log Out', style: TextStyle(color: AppColors.secondary)),
              ),
            ],
          ),
        ],
        iconTheme: const IconThemeData(color: AppColors.primary),
      ),
      body: Stack(
        children: [
          // Red Felt Background
          Positioned(
            top: -100, bottom: -100, left: -50, right: -50,
            child: Container(
              decoration: BoxDecoration(
                color: const Color(0xFF8B0000).withValues(alpha: 0.6), // Dark Red Felt
                borderRadius: BorderRadius.circular(400),
                border: Border.all(color: AppColors.primary.withValues(alpha: 0.1), width: 2),
              ),
            ),
          ),
          // GRAND STAKES Watermark
          Center(
            child: Opacity(
              opacity: 0.05,
              child: Text(
                "GRAND\nSTAKES",
                textAlign: TextAlign.center,
                style: Theme.of(context).textTheme.displayLarge?.copyWith(
                  fontSize: 80,
                  fontWeight: FontWeight.w900,
                  color: Colors.white,
                  height: 1.0,
                  fontStyle: FontStyle.italic,
                ),
              ),
            ),
          ),
          // Game Elements
          SafeArea(
            child: Column(
              children: [
                const SizedBox(height: 32),
                const Text("DEALER STANDS ON 17", style: TextStyle(color: AppColors.primary, fontSize: 10, letterSpacing: 2.0, fontWeight: FontWeight.bold)),
                const SizedBox(height: 24),
                
                // Dealer Area
                if (_dealerH.isNotEmpty)
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: _dealerH.asMap().entries.map((e) {
                      bool isHidden = !_gameOver && e.key == 0;
                      return Padding(
                        padding: const EdgeInsets.symmetric(horizontal: 4),
                        child: _PlayingCardMockup(card: e.value, isHidden: isHidden),
                      );
                    }).toList(),
                  ),
                
                const Spacer(),
                
                // Center Banner
                Container(
                  padding: const EdgeInsets.symmetric(horizontal: 32, vertical: 8),
                  decoration: BoxDecoration(
                    color: Colors.black.withValues(alpha: 0.4),
                    borderRadius: BorderRadius.circular(4),
                  ),
                  child: const Text("BLACKJACK PAYS 3 TO 2", style: TextStyle(color: AppColors.primary, fontSize: 12, letterSpacing: 2.0, fontWeight: FontWeight.bold, fontStyle: FontStyle.italic)),
                ),
                
                const Spacer(),
                
                // Player Area
                if (_splitH.isNotEmpty) ...[
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    children: [
                      // Hand 1
                      Container(
                        padding: const EdgeInsets.all(8),
                        decoration: BoxDecoration(border: Border.all(color: _activeHand == 1 ? AppColors.primary : Colors.transparent, width: 2), borderRadius: BorderRadius.circular(12)),
                        child: Row(children: _playerH.map((c) => Padding(padding: const EdgeInsets.symmetric(horizontal: 2), child: _PlayingCardMockup(card: c))).toList()),
                      ),
                      // Hand 2
                      Container(
                        padding: const EdgeInsets.all(8),
                        decoration: BoxDecoration(border: Border.all(color: _activeHand == 2 ? AppColors.primary : Colors.transparent, width: 2), borderRadius: BorderRadius.circular(12)),
                        child: Row(children: _splitH.map((c) => Padding(padding: const EdgeInsets.symmetric(horizontal: 2), child: _PlayingCardMockup(card: c))).toList()),
                      ),
                    ],
                  )
                ] else if (_playerH.isNotEmpty) ...[
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: _playerH.map((c) => Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 4),
                      child: _PlayingCardMockup(card: c),
                    )).toList(),
                  ),
                ],
                
                const SizedBox(height: 16),
                
                if (_message.isNotEmpty)
                  AnimatedScale(
                    scale: _message.contains("WIN") || _message.contains("BLACKJACK") ? 1.2 : 1.0,
                    duration: const Duration(milliseconds: 500),
                    curve: Curves.elasticOut,
                    child: Text(
                      _message,
                      style: Theme.of(context).textTheme.headlineMedium?.copyWith(
                        color: _message.contains("WIN") || _message.contains("BLACKJACK") ? Colors.amber : AppColors.primary,
                        shadows: _message.contains("WIN") || _message.contains("BLACKJACK") ? [const Shadow(color: Colors.amberAccent, blurRadius: 10)] : [],
                      ),
                    ),
                  ),
                
                const SizedBox(height: 24),
                
                // Controls
                if (!_isPlaying) ...[
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      _buildChip(10),
                      const SizedBox(width: 8),
                      _buildChip(50),
                      const SizedBox(width: 8),
                      _buildChip(100),
                      const SizedBox(width: 8),
                      _buildChip(500),
                    ],
                  ),
                  const SizedBox(height: 16),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      IconButton(icon: const Icon(Icons.clear, color: Colors.red), onPressed: () => setState(()=> _betAmount = 10)),
                      Text("TOTAL BET: \$$_betAmount", style: Theme.of(context).textTheme.headlineMedium?.copyWith(color: AppColors.primary)),
                    ],
                  ),
                  const SizedBox(height: 16),
                  InkWell(
                    onTap: _placeBet,
                    child: Ink(
                      padding: const EdgeInsets.symmetric(horizontal: 64, vertical: 16),
                      decoration: const BoxDecoration(
                        gradient: LinearGradient(
                          begin: Alignment.topLeft,
                          end: Alignment.bottomRight,
                          colors: [AppColors.primary, AppColors.primaryContainer],
                        ),
                      ),
                      child: Text("PLACE BET", style: Theme.of(context).textTheme.titleMedium?.copyWith(color: AppColors.surface, fontWeight: FontWeight.bold)),
                    ),
                  )
                ] else ...[
                  Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 16),
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                      children: [
                        _buildActionButton("FOLD", _fold),
                        _buildActionButton("STAND", _stand),
                        _buildActionButton("HIT", _hit, isPrimary: true),
                        _buildActionButton("DOUBLE", _doubleDown),
                        _buildActionButton("SPLIT", _split),
                      ],
                    ),
                  ),
                ],
                const SizedBox(height: 32),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _PlayingCardMockup extends StatefulWidget {
  final PlayingCard? card;
  final bool isHidden;

  const _PlayingCardMockup({Key? key, this.card, this.isHidden = false}) : super(key: key);

  @override
  State<_PlayingCardMockup> createState() => _PlayingCardMockupState();
}

class _PlayingCardMockupState extends State<_PlayingCardMockup> with SingleTickerProviderStateMixin {
  late AnimationController _ctrl;
  late Animation<Offset> _slideAnim;

  @override
  void initState() {
    super.initState();
    _ctrl = AnimationController(vsync: this, duration: const Duration(milliseconds: 600));
    _slideAnim = Tween<Offset>(begin: const Offset(0, -6), end: Offset.zero).animate(CurvedAnimation(parent: _ctrl, curve: Curves.easeOutBack)); // Salto de la carta con 3D
    _ctrl.forward();
  }

  @override
  void dispose() {
    _ctrl.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SlideTransition(
      position: _slideAnim,
      child: AnimatedBuilder(
        animation: _ctrl,
        builder: (context, child) {
          return Transform(
            transform: Matrix4.identity()
              ..setEntry(3, 2, 0.001)
              ..rotateX((1.0 - _ctrl.value) * 1.5) // Voltea la carta en 3D mientras cae
              ..rotateZ(((widget.card?.hashCode ?? 0) % 10 - 5) * 0.01), // Ligera inclinación natural
            alignment: Alignment.center,
            child: child,
          );
        },
        child: _buildCard(),
      ),
    );
  }
  
  Widget _buildCard() {
    if (widget.isHidden || widget.card == null) {
      return Container(
        width: 80,
        height: 110,
        decoration: BoxDecoration(
          color: const Color(0xFF382A2A), // Dark brown back
          borderRadius: BorderRadius.circular(6),
          border: Border.all(color: AppColors.primary.withValues(alpha: 0.5), width: 1.5),
          boxShadow: const [BoxShadow(color: Colors.black54, blurRadius: 4, offset: Offset(2, 2))],
        ),
        alignment: Alignment.center,
        child: const Icon(Icons.diamond_outlined, color: AppColors.primary, size: 32),
      );
    }
    
    return Container(
      width: 80,
      height: 110,
      decoration: BoxDecoration(
        color: const Color(0xFFEAEAEA),
        borderRadius: BorderRadius.circular(6),
        boxShadow: const [BoxShadow(color: Colors.black54, blurRadius: 4, offset: Offset(2, 2))],
      ),
      padding: const EdgeInsets.all(6),
      child: Stack(
        children: [
          Positioned(
            top: 0, left: 0,
            child: Column(
              children: [
                Text(widget.card!.label, style: TextStyle(color: widget.card!.isRed ? const Color(0xFFD32F2F) : Colors.black87, fontSize: 18, fontWeight: FontWeight.w900)),
                Text(widget.card!.suitSymbol, style: TextStyle(color: widget.card!.isRed ? const Color(0xFFFFCDD2) : Colors.black38, fontSize: 12)),
              ],
            ),
          ),
          Center(
            child: Text(
              widget.card!.suitSymbol, 
              style: TextStyle(
                color: widget.card!.isRed ? const Color(0xFFD32F2F).withValues(alpha: 0.15) : Colors.black.withValues(alpha: 0.1), 
                fontSize: 60
              )
            ),
          ),
          Positioned(
            bottom: 0, right: 0,
            child: RotatedBox(
              quarterTurns: 2,
              child: Column(
                children: [
                  Text(widget.card!.label, style: TextStyle(color: widget.card!.isRed ? const Color(0xFFD32F2F) : Colors.black87, fontSize: 18, fontWeight: FontWeight.w900)),
                  Text(widget.card!.suitSymbol, style: TextStyle(color: widget.card!.isRed ? const Color(0xFFFFCDD2) : Colors.black38, fontSize: 12)),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: app/src/main/kotlin/com/grandstakes/ui/games/BlackjackScreen.kt
package com.grandstakes.ui.games

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Diamond
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.grandstakes.logic.Deck
import com.grandstakes.logic.PlayingCard
import com.grandstakes.ui.main.LobbyViewModel
import com.grandstakes.ui.components.GrandStakesButton
import com.grandstakes.ui.theme.GoldPrimary
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BlackjackScreen(
    onNavigateBack: () -> Unit,
    viewModel: LobbyViewModel = hiltViewModel()
) {
    val scope = rememberCoroutineScope()
    val deck = remember { Deck() }
    
    // Game State
    var userHand by remember { mutableStateOf(listOf<PlayingCard>()) }
    var dealerHand by remember { mutableStateOf(listOf<PlayingCard>()) }
    var isPlaying by remember { mutableStateOf(false) }
    var gameOver by remember { mutableStateOf(false) }
    var isProcessing by remember { mutableStateOf(false) }
    var message by remember { mutableStateOf("PLACE YOUR BET") }
    var betAmount by remember { mutableIntStateOf(100) }
    var activeBet by remember { mutableIntStateOf(0) }

    val user by viewModel.currentUser.collectAsState()
    val balance = user?.balance ?: 0

    fun calculateHand(hand: List<PlayingCard>): Int {
        if (hand.isEmpty()) return 0
        var total = hand.sumOf { it.blackjackValue }
        var aces = hand.count { it.rank.label == "A" }
        while (total > 21 && aces > 0) {
            total -= 10
            aces--
        }
        return total
    }

    fun endGame() {
        gameOver = true
        isPlaying = false
        val uScore = calculateHand(userHand)
        val dScore = calculateHand(dealerHand)

        if (uScore > 21) {
            message = "BUST! YOU LOSE."
        } else {
            if (uScore == 21 && userHand.size == 2 && !(dScore == 21 && dealerHand.size == 2)) {
                message = "BLACKJACK! WIN 3:2"
                viewModel.updateBalance((activeBet * 2.5).toInt(), "Blackjack Win")
            } else if (dScore > 21 || uScore > dScore) {
                message = "YOU WIN!"
                viewModel.updateBalance(activeBet * 2, "Blackjack Win")
            } else if (uScore < dScore) {
                message = "DEALER WINS."
            } else {
                message = "PUSH (TIE)."
                viewModel.updateBalance(activeBet, "Blackjack Push")
            }
        }
    }

    fun deal() {
        if (isProcessing) return
        if (balance < betAmount) {
            message = "INSUFFICIENT CREDIT"
            return
        }

        scope.launch {
            isProcessing = true
            message = "DEALING..."
            viewModel.updateBalance(-betAmount, "Blackjack Bet")
            activeBet = betAmount
            deck.reset()
            userHand = emptyList()
            dealerHand = emptyList()
            gameOver = false
            isPlaying = true

            // Animation sequence
            userHand = userHand + deck.draw()
            delay(400)
            dealerHand = dealerHand + deck.draw()
            delay(400)
            userHand = userHand + deck.draw()
            delay(400)
            dealerHand = dealerHand + deck.draw()
            delay(400)

            message = ""
            if (calculateHand(userHand) == 21) {
                endGame()
            }
            isProcessing = false
        }
    }

    fun hit() {
        if (isProcessing || gameOver) return
        scope.launch {
            isProcessing = true
            userHand = userHand + deck.draw()
            if (calculateHand(userHand) > 21) {
                endGame()
            }
            isProcessing = false
        }
    }

    fun stand() {
        if (isProcessing || gameOver) return
        scope.launch {
            isProcessing = true
            while (calculateHand(dealerHand) < 17) {
                message = "DEALER DRAWS..."
                dealerHand = dealerHand + deck.draw()
                delay(600)
            }
            message = ""
            endGame()
            isProcessing = false
        }
    }

    Scaffold(
        containerColor = Color.Black,
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("GRAND BLACKJACK", style = MaterialTheme.typography.labelSmall.copy(color = GoldPrimary, letterSpacing = 2.sp)) },
                navigationIcon = {
                    TextButton(onClick = onNavigateBack) {
                        Text("EXIT", color = GoldPrimary)
                    }
                },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = Color.Black)
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                "DEALER STANDS ON 17   •   PAYS 3:2",
                style = MaterialTheme.typography.labelSmall.copy(color = GoldPrimary.copy(alpha = 0.5f), letterSpacing = 1.sp)
            )
            
            Spacer(modifier = Modifier.weight(1f))

            // Dealer Area
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                BJHandView(hand = dealerHand, isDealer = true, isHidden = !gameOver && isPlaying)
                Spacer(modifier = Modifier.height(8.dp))
                if (dealerHand.isNotEmpty()) {
                    Text(
                        if (!gameOver && isPlaying) "?" else calculateHand(dealerHand).toString(),
                        style = MaterialTheme.typography.titleMedium.copy(color = Color.White.copy(alpha = 0.5f))
                    )
                }
            }

            Spacer(modifier = Modifier.weight(1f))

            // Message Area
            Text(
                message.uppercase(),
                style = MaterialTheme.typography.headlineSmall.copy(
                    color = GoldPrimary,
                    fontWeight = FontWeight.Bold,
                    textAlign = TextAlign.Center
                ),
                modifier = Modifier.height(40.dp)
            )

            Spacer(modifier = Modifier.weight(1f))

            // Player Area
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                if (userHand.isNotEmpty()) {
                    Text(
                        calculateHand(userHand).toString(),
                        style = MaterialTheme.typography.titleMedium.copy(color = GoldPrimary)
                    )
                }
                Spacer(modifier = Modifier.height(8.dp))
                BJHandView(hand = userHand)
            }

            Spacer(modifier = Modifier.weight(1f))

            // Controls
            if (!isPlaying || (gameOver && !isProcessing)) {
                BetControls(
                    betAmount = betAmount,
                    balance = balance,
                    onBetChange = { betAmount = it },
                    onDeal = { deal() },
                    disabled = isProcessing
                )
            } else {
                ActionControls(
                    onHit = { hit() },
                    onStand = { stand() },
                    disabled = isProcessing
                )
            }
        }
    }
}

@Composable
fun BJHandView(hand: List<PlayingCard>, isDealer: Boolean = false, isHidden: Boolean = false) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(120.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically
    ) {
        hand.forEachIndexed { index, card ->
            key(card.id) {
                BJCardView(
                    card = card,
                    isHidden = isDealer && isHidden && index == 0
                )
            }
        }
    }
}

@Composable
fun BJCardView(card: PlayingCard, isHidden: Boolean) {
    Surface(
        modifier = Modifier
            .padding(horizontal = 4.dp)
            .size(width = 70.dp, height = 100.dp),
        color = if (isHidden) Color(0xFF1E1E1E) else Color.White,
        shape = RoundedCornerShape(4.dp),
        border = BorderStroke(1.dp, GoldPrimary.copy(alpha = 0.2f))
    ) {
        if (isHidden) {
            Box(contentAlignment = Alignment.Center) {
                Icon(Icons.Default.Diamond, contentDescription = null, tint = GoldPrimary, modifier = Modifier.size(24.dp))
            }
        } else {
            val color = if (card.suit.isRed) Color(0xFFFF5252) else Color.Black
            Column(modifier = Modifier.padding(6.dp)) {
                Text(card.rank.label, color = color, fontWeight = FontWeight.Black, fontSize = 16.sp)
                Text(card.suit.symbol, color = color, fontSize = 12.sp)
                Spacer(modifier = Modifier.weight(1f))
                Box(modifier = Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                    Icon(Icons.Default.Diamond, contentDescription = null, tint = color.copy(alpha = 0.05f), modifier = Modifier.size(32.dp))
                }
                Spacer(modifier = Modifier.weight(1f))
                Column(modifier = Modifier.rotate(180f).align(Alignment.End)) {
                    Text(card.rank.label, color = color, fontWeight = FontWeight.Black, fontSize = 16.sp)
                    Text(card.suit.symbol, color = color, fontSize = 12.sp)
                }
            }
        }
    }
}

@Composable
fun BetControls(
    betAmount: Int,
    balance: Int,
    onBetChange: (Int) -> Unit,
    onDeal: () -> Unit,
    disabled: Boolean
) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = { if (betAmount > 10) onBetChange(betAmount - 10) }, enabled = !disabled) {
                Icon(Icons.Default.Remove, contentDescription = null, tint = GoldPrimary)
            }
            Text(
                "$${betAmount}",
                style = MaterialTheme.typography.displayMedium.copy(color = Color.White, fontSize = 42.sp)
            )
            IconButton(onClick = { if (betAmount + 10 <= balance) onBetChange(betAmount + 10) }, enabled = !disabled) {
                Icon(Icons.Default.Add, contentDescription = null, tint = GoldPrimary)
            }
        }
        Spacer(modifier = Modifier.height(16.dp))
        GrandStakesButton(
            text = "PLACE BET",
            onClick = onDeal,
            enabled = !disabled && balance >= betAmount,
            modifier = Modifier.width(220.dp)
        )
    }
}

@Composable
fun ActionControls(onHit: () -> Unit, onStand: () -> Unit, disabled: Boolean) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = 32.dp),
        horizontalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        Button(
            onClick = onStand,
            modifier = Modifier.weight(1f).height(56.dp),
            enabled = !disabled,
            colors = ButtonDefaults.buttonColors(containerColor = Color.Transparent),
            border = BorderStroke(1.dp, GoldPrimary),
            shape = RoundedCornerShape(4.dp)
        ) {
            Text("STAND", color = GoldPrimary, fontWeight = FontWeight.Bold)
        }
        Button(
            onClick = onHit,
            modifier = Modifier.weight(1f).height(56.dp),
            enabled = !disabled,
            colors = ButtonDefaults.buttonColors(containerColor = GoldPrimary),
            shape = RoundedCornerShape(4.dp)
        ) {
            Text("HIT", color = Color.Black, fontWeight = FontWeight.Bold)
        }
    }
}
